### PR TITLE
Fix secret controller racetest issues

### PR DESCRIPTION
**Pull request by john-a-joyce**
This change fixes racetest failure that are flagged
in the packages that consume this library.

### CODE REVIEW

File: pkg/kube/secretcontroller/secretcontroller.go [hunk 1]
```diff
@@ -242,11 +245,13 @@ func (c *Controller) addMemberCluster(secretName string, s *corev1.Secret) {
 			log.Infof("Cluster %s in the secret %s in namespace %s already exists",
 				clusterID, c.cs.remoteClusters[clusterID].secretName, s.ObjectMeta.Namespace)
 		}
+		c.Unlock()
```
Comment (nmittler) [inline pkg/kube/secretcontroller/secretcontroller.go]: This doesn't look right. You have cases above that `continue`, which would miss this unlock call. Maybe you should move all of this to a function and use `defer c.Unlock`?

File: pkg/kube/secretcontroller/secretcontroller.go [hunk 2]
```diff
@@ -257,5 +262,6 @@ func (c *Controller) deleteMemberCluster(secretName string) {
 			delete(c.cs.remoteClusters, clusterID)
 		}
 	}
+	c.Unlock()
```
Comment (nmittler) [inline pkg/kube/secretcontroller/secretcontroller.go]: Use `defer c.Unlock()` immediately after `c.Lock()`

File: pkg/kube/secretcontroller/secretcontroller.go (hunk 1 above)
Comment (john-a-joyce) [reply pkg/kube/secretcontroller/secretcontroller.go]: Yeah - I missed the continue in my haste - thanks for catching.   For now I am just checking if this has any influence on the racetest failures introduced in the pilot multicluster code.  I don't think it does but don't have a good way to run the reacetest locally.   I am likely to abandon this PR but changed per your suggestion anyway.

**Comment by baodongli** [comment]
The added lock doesn't seem to be used at runtime, but it's added to address a race condition between the secret controller thread and the test thread. Apparently the test thread accesses data that is directly under control of the secret controller thread. Not sure how to revise the test code so that the lock doesn't have to be added to work with the test code.

**Comment by john-a-joyce** [comment]
This PR is not needed the proper fix is in PR #9267

**Comment by baodongli** [comment]
One way to refactor this code to incorporate a lock and make the user's life a bit easier is to make the add callback to return an interface and save it in the remote cluster data. The user's code doesn't have to maintain a list of remote clusters any more. This list is maintained in the secret controller and access to it is lock protected. The secret controller needs to provide query routine to access them. This requires both the secret controller and the user's modules (pilot and mixer) to be refactored.

**Pull request closed without merging**